File: Feb-2023/C++/FindDuplicatesSubtrees.cxx
// 652. Find Duplicate Subtrees

/**

 * Definition for a binary tree node.

 * struct TreeNode {

 *     int val;

 *     TreeNode *left;

 *     TreeNode *right;

 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}

 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}

 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}

 * };

 */

class Solution {

	public:	string turnintosubstring(TreeNode * root, vector<TreeNode*> & res, unordered_map<string, int> & map) {

		// add null values too, to make sure there they have the right structure even if the strings match

		if (root == nullptr) return "NULL";

		// create a string starting from the root 

		string s = to_string(root->val) + "," + turnintosubstring(root->left, res, map) + "," + turnintosubstring(root->right, res, map);

		// if theres one duplicage

		if (map[s] == 1) {

			res.push_back(root);

		} 

          // no else for duplicates

		 map[s]++;		

		return s;

	}

	vector<TreeNode*> findDuplicateSubtrees(TreeNode * root) {

		vector<TreeNode*> res;

		if (root == nullptr) return res;

		unordered_map<string, int> map;

		turnintosubstring(root, res, map);

		return res;

	}

};

File: Feb-2023/Java/FindDuplicatesSubtrees.java
// 652. Find Duplicate Subtrees
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public String checkstring(TreeNode root,HashMap<String,Integer> map, List<TreeNode> res){
        String s = "";
        // Null check added to match schema
        if(root==null) return "NULL";
        // Starting from the root, keep track of all values from a particular node
        s = String.valueOf(root.val)+","+checkstring(root.left,map,res)+","+checkstring(root.right,map,res);
       // Only add the root if duplicate is found once
        if(map.containsKey(s) && map.get(s)==1){
            res.add(root);
         }
        // Add the count to make sure to not add it if it's a duplicate in the next recursion call in the above code
        map.put(s,map.getOrDefault(s,0)+1);
        return s;
     }
    
    public List<TreeNode> findDuplicateSubtrees(TreeNode root) {
        List<TreeNode> res = new LinkedList<>();
        HashMap<String,Integer> map = new HashMap<>();
        if(root==null) return res;
        checkstring(root,map,res);
        return res;
    }
}
